Frontend: Add build:stats for analysing bundles

jackw opened the pull request:
<!--

Thank you for sending a pull request! Here are some tips:

1. If this is your first time, please read our contribution guide at https://github.com/grafana/grafana/blob/main/CONTRIBUTING.md

2. Ensure you include and run the appropriate tests as part of your Pull Request.

3. In a new feature or configuration option, an update to the documentation is necessary. Everything related to the documentation is under the docs folder in the root of the repository.

4. If the Pull Request is a work in progress, make use of GitHub's "Draft PR" feature and mark it as such.

5. If you can not merge your Pull Request due to a merge conflict, Rebase it. This gets it in sync with the main branch.

6. Name your PR as "<FeatureArea>: Describe your change", e.g. Alerting: Prevent race condition. If it's a fix or feature relevant for the changelog describe the user impact in the title. The PR title is used to auto-generate the changelog for issues marked with the "add to changelog" label.

7. If your PR content should be added to the What's New document for the next major or minor release, add the **add to what's new** label to your PR. Note that you should add this label to the main PR that introduces the feature; do not add this label to smaller PRs for the feature.

-->

**What is this feature?**

This PR adds a `build:stats` script and an additional webpack.stats config file that by default will append the `webpack-bundle-analyser` to the webpack production config.

Additional env vars can be passed to enable the following:

- `yarn build:stats --env namedChunks`: replaces the hash names in production assets with a human readable name
- `yarn build:stats --env doctor`: adds the webpack rsdoctor plugin alongside the bundle-analyser plugin

these can be combined `yarn build:stats --env namedChunks --env doctor` if need be.

Rsdoctor contains a [collection of features](https://rsdoctor.dev/guide/start/intro#-features) that appear to work better than [statoscope webpack plugin](https://github.com/statoscope/statoscope). Below are a few screengrabs for reviewers.

| | | 
| --- | --- |
| <img width="1792" alt="Screenshot 2024-10-15 at 15 04 31" src="https://github.com/user-attachments/assets/d2e9609f-d0da-46b8-a2be-13331234cca1"> | <img width="1792" alt="Screenshot 2024-10-15 at 15 04 31" src="https://github.com/user-attachments/assets/94fb260a-2a14-456c-b644-3dd4fb4d0dc1"> | 
| <img width="1792" alt="Screenshot 2024-10-15 at 15 07 18" src="https://github.com/user-attachments/assets/17eee70f-19fb-40f8-84bd-fb45e1f18591"> | <img width="1792" alt="Screenshot 2024-10-15 at 15 42 10" src="https://github.com/user-attachments/assets/2790cc7c-9849-4edb-8601-c544db55f246"> | 

**Why do we need this feature?**

This should make it easier for Grafana contributors to analyse and reduce frontend asset sizes.

**Who is this feature for?**

Performance minded folk.

**Which issue(s) does this PR fix?**:

<!--

- Automatically closes linked issue when the Pull Request is merged.

Usage: "Fixes #<issue number>", or "Fixes (paste link of issue)"

-->

Fixes #

**Special notes for your reviewer:**

Please check that:
- [ ] It works as expected from a user's perspective.
- [ ] If this is a pre-GA feature, it is behind a feature toggle.
- [ ] The docs are updated, and if this is a [notable improvement](https://grafana.com/docs/writers-toolkit/contribute/release-notes/#how-to-determine-if-content-belongs-in-whats-new), it's added to our [What's New](https://grafana.com/docs/writers-toolkit/contribute/release-notes/) doc.

jackw committed:
build(webpack): add a stats config and build:stats script for analysing bundles locally

jackw committed:
chore(yarn): dedupe lock file

leeoniya commented:
this is great! some comments:

the multiple/duplicate package versions is sweet and will help with keeping that tidy.

> and an additional webpack.stats config file that disables chunk hash names for legibility

for me it actually reduces legibility with hashed names i can ignore them more easily than names that look like they're modules. it increases the noise in by eyesbrain :)

the bundle inspection is unfortunately only half screen with no option to collapse the header or make it full screen. with so much nesting and dependencies it's hard to get a as much legibility/value out of it as with bare/fullpage `webpack-bundle-analyzer`.

maybe we can also add a more simple `build:bundalyze` script that only uses `webpack-bundle-analyzer`?

![image](https://github.com/user-attachments/assets/78c30e33-57a3-4eb3-ac82-adc8b155e4ca)

jackw commented:
>for me this actually reduces legibility. with hashed names i can ignore them more easily than names that look like they're modules :) another important factor is that i can relate the local bundle names to those is see in actual production instances in DevTools/Network and Code coverage panels

I see. I actually find it easier to navigate to the file I want to check but I'm not overly precious on this so can revert or maybe set it to only apply based on an env var?

>maybe we can also add a more simple build:bundalyze script that only uses webpack-bundle-analyzer

Hmm... maybe rather than using rsdoctor we'd do better to use both the [statoscope webpack plugin](https://statoscope.tech/#&hash=6b5d7403b35900e465b8) and the bundle-analyzer plugin in the stats script? This gives us pretty much everything rs-doctor provides but opens separate tabs for each plugin? WDYT?

jackw committed:
feat(webpack): use bundle analyser by default, env vars for additional plugins

jackw commented:
@leeoniya I've made the following changes (and updated the PR description accordingly)...

By default this `build:stats` will only add the bundle-analyser plugin. I've added a couple of env vars to the mix so devs can swap out the hashed names for human readable ones and add the rsdoctor plugin for analysing duplicate npm packages and reasons for why code is included in bundles.

jackw commented:
/deploy-to-hg
